File: day01/src/main.rs
use std::{fs, iter::zip};


fn parse_lines(lines: &Vec<&str>) -> (Vec<i32>, Vec<i32>) {
    let mut left: Vec<i32> = vec![];
    let mut right: Vec<i32> = vec![];
    for line in lines.iter() {
        if line.is_empty() {
            continue;
        }
        let mut line = line.split_whitespace();
        left.push(line.next().unwrap_or("0").parse().unwrap_or(0));
        right.push(line.next().unwrap_or("0").parse().unwrap_or(0));
    }
    (left, right)
}

fn part1(lines: &Vec<&str>, expected: Option<i32>) -> i32 {
    // Implementation here
    let (mut left, mut right) = parse_lines(lines);
    left.sort();
    right.sort();
    let total = zip(left, right).map(|(l, r)| (r - l).abs()).fold(0, |acc, e| acc + e);

    if let Some(exp) = expected {
        println!("Part 1 Calculated: {total} expected: {exp}");
        // assert_eq!(exp, total);
    }
    total
}

fn part2(lines: &Vec<&str>, expected: Option<i32>) -> i32  {
    // Implementation here
    let (mut left, mut right) = parse_lines(lines);
    let mut total = 0;
    for l in left {
        total += l * right.iter().filter(|r| r == &&l).count() as i32;
    }
    if let Some(exp) = expected {
        println!("Part 2 Calculated: {total} expected: {exp}");
        // assert_eq!(exp, total);
    }
    total
}

fn process_file(path: &str, exp1: Option<i32>, exp2: Option<i32>) -> (i32, i32) {
    let input = fs::read_to_string(path).unwrap();
    let lines: Vec<&str> = input.split('\n').into_iter().map(|l| {l.trim()}).collect();
    let part1_result = part1(&lines, exp1);
    let part2_result = part2(&lines, exp2);
    (part1_result, part2_result)
}

fn main() {
    let example_part1_expected = 11;
    let example_part2_expected = 31;
    let (ex1, _) = process_file("example1", Some(example_part1_expected), None);
    let (_, ex2) = process_file("example1", None, Some(example_part2_expected));
    println!("Example 1: {ex1} 2: {ex2}");
    let (p1, p2) = process_file("input", None, None);
    println!("Input: part 1: {p1} part 2: {p2}");
}

File: day01/main.py
# https://adventofcode.com/2024/day/1

import collections
import os
from typing import List

def parse_lines(lines: List[str]) -> tuple[List[int], List[int]]:
    left = []
    right = []
    for line in lines:
        l, r = line.split()
        left.append(int(l))
        right.append(int(r))
    return left, right

def part1(lines: List[str], expected: int | None = None) -> int:
    total = 0
    # Start of implementation #
    left, right = parse_lines(lines)
    left.sort()
    right.sort()
    for l, r in zip(left, right):
        total += abs(r - l)
    # End of implementation #
    if expected is not None:
        assert total == expected
    return total


def part2(lines: List[str], expected: int | None = None) -> int:
    # Start of implementation #
    left, right = parse_lines(lines)
    counts = collections.defaultdict(lambda: 0)
    for r in right:
        counts[r] += 1
    total = 0
    for l in left:
        total += l * counts[l]
    # End of implementation #
    if expected is not None:
        assert total == expected
    return total


def process_file(filename: str, expected_part1: int | None = None, expected_part2: int | None = None):
    path = os.path.join(os.path.dirname(__file__), filename)
    with open(path, "r") as input_file:
        lines = input_file.readlines()

    p1 = part1(lines, expected_part1)
    p2 = part2(lines, expected_part2)
    return p1, p2


if __name__ == "__main__":
    example_part1_expected = 11
    example_part2_expected = 31
    ex1, _ = process_file("example1", example_part1_expected, None)
    _, ex2 = process_file("example1", None, example_part2_expected)
    print("Example:")
    print(f"Part 1 expected {example_part1_expected} got {ex1}: {example_part1_expected == ex1}")
    print(f"Part 2 expected {example_part2_expected} got {ex2}: {example_part2_expected == ex2}")

    p1, p2 = process_file("input")
    print("Input:")
    print(f"Part 1: {p1}")
    print(f"Part 2: {p2}")